x-pack/plugins/gis/public/components/widget_overlay/attribution_control/view.js

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { Fragment } from 'react';
+import _ from 'lodash';
+import {
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiPanel,
+  EuiLink,
+} from '@elastic/eui';
+
+
+export class AttributionControl  extends React.Component {
+
+  constructor() {
+    super();
+    this.state = {
+      uniqueAttributions: []
+    };
+  }
+
+  componentDidMount() {
+    this._isMouunted = true;
+  }
+
+  componentWillUnmount() {
+    this._isMouunted = false;
+  }
+
+  _syncMbMapWithAttribution = async () => {
+
+    const attributionPromises = this.props.layerList.map(layer => {
+      return layer.getAttributions();
+    });
+    const attributions = await Promise.all(attributionPromises);
+    if (!this._isMouunted) {
+      return;
+    }
+
+    const uniqueAttributions = [];
```

Comment:
Can you move all of this logic to a redux selector and then just pass in the attributions to the component in the connector? We need to avoid having side effects like this in the render method

Reply:
If you're OK with keeping this method async, it will be tough doing this in a selector, since selectors need to be synchronous.

The async-iness introduces the need to store this in internal component-state (since we can only bind components to their properties synchronously).

but yes, we can move this to a selector, but only if we remove the async signature.

Reply:
If there are reasons to keep it async then you are right, it would not work as a selector

`this._syncMbMapWithAttribution` should not be triggered by `render`. You will need to trigger it from `getDerivedStateFromProps` instead.